### ISSUE TITLE
The role of the VAE model

### ISSUE THREAD

**Issue by edikedik** [LirongWu/Prompt-DDG#3]
Dear authors,

I evaluated Prompt-DDG on the [HER2 dataset from the GearBind model](https://github.com/DeepGraphLearning/GearBind/tree/main/data) and was highly impressed by the results:

```
   N_mut  Count  PearsonR  SpearmanR
0      2     18      0.54       0.37
1      3      4      0.98       1.00
2      4     10      0.74       0.65
3      5     24      0.71       0.48
4      6     52      0.61       0.46
5      7     75      0.56       0.53
6      8     87      0.52       0.48
7      9     70      0.36       0.38
8     10     41      0.35       0.42
9     11     24      0.13       0.17
10    12     11     -0.15      -0.05
11    13      2     -1.00      -1.00
12   All    419      0.51       0.53
   N_mut  Count  PearsonR  SpearmanR
0      2     18      0.58       0.43
1      3      4      0.99       0.80
2      4     10      0.67       0.62
3      5     24      0.72       0.48
4      6     52      0.63       0.54
5      7     75      0.68       0.62
6      8     87      0.51       0.51
7      9     70      0.51       0.51
8     10     41      0.30       0.31
9     11     24      0.20       0.26
10    12     11      0.40       0.39
11    13      2      1.00       1.00
12   All    419      0.58       0.61
   N_mut  Count  PearsonR  SpearmanR
0      2     18      0.69       0.56
1      3      4      0.95       0.80
2      4     10      0.69       0.58
3      5     24      0.70       0.52
4      6     52      0.61       0.51
5      7     75      0.51       0.49
6      8     87      0.53       0.52
7      9     70      0.42       0.44
8     10     41      0.32       0.32
9     11     24     -0.00       0.01
10    12     11      0.18       0.21
11    13      2      1.00       1.00
12   All    419      0.52       0.54
```
Above, each trained model was evaluated separately.

It motivated me to dig deeper into the code and its connection to the published work. What I found perplexing is the precise role of the VQ-VAE model for the ddG prediction. In the `train.py` script this model is initialized anew and trained on a single batch. This implies that no extensive pretraining is required. Moreover, I carried out the pretraining of the VAE and confirmed that it does not affect the ddG predictor's output, which begs the question: what is the VAE's (`Codebook`) role? Is it required or rather serves as a placeholder?

Sincerely,
Ivan

**Comment by LirongWu** [comment]
Dear Ivan.

Whether VQ-VAE is pre-trained and used in train.py depends on the hyperparameter _args.pre_epoch_. When _args.pre_epoch==0_, it does not pre-train or use VQ-VAE in subsequent training and inference. Thus, VQ-VAE is not a placeholder, it is required but it is allowed to be masked by controling the hyperparameter. When args.pre_epoch!=0, it will pre-train VQ-VAE and use it in subsequent training and inference.

Sincerely.
Lirong

**Comment by edikedik** [comment]
Thanks, Lirong. Of course, evidently, the VQ-VAE model is _used_ for inference. So my question pertains to whether it has any actual _impact_ on ddG predictions. Because, you see, when one carries out inference on SKEMPI in a suggested manner (`python -B train.py --ckpt_path ../trained_models/ddg_model.ckpt`), no existing VQ-VAE weights are used, and a _new_ VQ-VAE model is initialized and _trained on a single batch_ -- undoubtedly insufficient to learn a meaningful codebook encompassing possible microenvironment states -- which still leads to the CV performance reported in a paper. In other words, the predictions seem _indifferent_ to the embeddings pulled out from the VQ-VAE. I'm finding this surprising. Could you please clarify?

Cheers

**Comment by LirongWu** [comment]
You raise a good question.The output of VQ-VAE (codebook) is used in both stages (1) training DDG predictor and (2) inference using DDG predictor, but it plays a more important role in the former stage, that provides microenvironmental differences as prompts (guidelines) for training DDG predictor.

As for the inference in stage (2), we were previously surprised to find that the output of the VQ-VAE has very little effect on the DDG prediction during the inference process (as you have found), and there are two potential causes behind this: (1) the order-of-magnitude difference between the prompt embedding **p** and the residue embedding **x** makes **p** be viewed as noise to the inputs by the neural network, and (2) as a result of the prompt-guided training, the DDG predictor is already microenvironment-aware to some extent, which makes the additional precise microenvironment information of little benefit.

Therefore, as you can see, we only provided pre-training weights for the DDG_model and not for the VQ-VAE, as we believe that the latter is *helpful, albeit limited*, to the inference process. However, despite this, we still need the embedding of the VQ-VAE input as a placeholder to ensure consistency of training and inference.

**Comment by paoslaos** [comment]
Hey @edikedik (and @LirongWu ) can I ask you how you did encode the HER2 mutations in the data loader? The example from the RBD only covers single mutants but your data set has multiple mutations.

Thanks
Sincerely,
P.
